types/json8-patch/index.d.ts

```diff
@@ -59,23 +60,215 @@ export type JsonPatchOperation = AddOperation | RemoveOperation | ReplaceOperati
+ */
+export function apply<T>(doc: T, patch: JsonPatch, options?: { reversible: true }): ApplyResultWithRevert<T>;
+/**
+ * Alias for the `apply()` method.
+ */
+export function patch<T>(doc: T, patch: JsonPatch, options?: { reversible: false }): PatchResult<T>;
+/**
+ * Alias for the `apply()` method.
+ */
+export function patch<T>(doc: T, patch: JsonPatch, options?: { reversible: true }): PatchResultWithRevert<T>;
+/**
+ * ```javascript
+ * ooPatch.diff(true, false);
+ * // [{"op": "replace", "path": "", "value": "false"}]
+ *
+ * ooPatch.diff([], []);
+ * // []
+ *
+ * ooPatch.diff({}, { foo: "bar" });
+ * // [{"op": "add", "path": "/foo", "value": "bar"}]
+ * ```
+ *
+ * @returns A diff in the form of a JSON Patch for 2 JSON values.
+ */
+export function diff(doc1: unknown, doc2: unknown): JsonPatch;
+/**
+ * If the patch or apply method is called with a third argument `{reversible: true}` it will return an additional value in the form of a revert property.
+ *
+ * The revert object can be used to revert a patch on a document.
+ *
+ * ```javascript
+ * // apply the patch with the reversible option
+ * var applyResult = ooPatch.apply(doc, patch, { reversible: true });
+ * doc = applyResult.doc;
+ *
+ * // revert the patch
+ * doc = ooPatch.revert(doc, applyResult.revert).doc;
+ * // doc is strictly identical to the original
+ * ```
+ *
+ * See also the `buildRevertPatch()` function which offers more flexibility.
+ */
+export function revert<T>(doc: T, revertPatch: JsonPatch): ApplyResult<T>;
+/**
+ * This method only check for JSON Patch semantic.
+ *
+ * ```javascript
+ * ooPatch.valid({})  // false
+ * ooPatch.valid([{}] // false
+ * ooPatch.valid([{op: "foo", path: null, value: undefined}]) // false
+ * ooPatch.valid([{op: "add", path: "/foo"}]) // false
+ *
+ * ooPatch.valid([]) // true
+ * ooPatch.valid([{op: "add", path: "/foo", value: "bar"}]) // true
+ * ```
+ *
+ * @returns `true` if the patch is valid, `false` otherwise.
+ */
+export function valid(patch: JsonPatch): boolean;
+/**
+ * Builds a valid JSON Patch from the result of a reversible `apply` operation.
+ * You can then use this patch with `apply()` method to revert a previously applied patch.
+ *
+ * Because `buildRevertPatch` + `apply` offers more flexibility over `revert` it is preferred.
+ *
+ * - use pack/unpack with the result of `buildRevertPatch` making it ideal for storage or transport
+ * - reverse a revert (and so on...) with `{reversible: true}`
+ * - `diff` between reverts
+ * - merge multiple reverts into one
+ * - rebase reverts
+ *
+ * @param revert The revert value from the `apply()` or `patch()` method with `{reversible: true}`
+ * @returns JSON Patch
+ */
+export function buildRevertPatch(revert: JsonPatch): JsonPatch;
 
-export function diff(doc1: any, doc2: any): JsonPatch;
+export interface OperationResult<T> {
+    /**
+     * The patched document
+     */
+    doc: T;
```

Comment:
same as above, I don't think we can guarantee what this type `T` will be for the `doc` property. `previous` having type `T` is probably ok, since that is supposed to be the input doc and not the result of some operation or patch.

Reply:
The `previous` wouldn't work too. The previous value is the the `T`, the subset of `T` or the value of any property in the object. I would replace it with what I suggested in my previous comment or roll back to any / unknown.